Improve RegexOptions.Compiled codegen for ignoring case (in particular InvariantCulture)

stephentoub opened the pull request:
Improved RegexOptions.Compiled code gen, in particular for IgnoreCase

Several changes:
- Improve the code generation for finding the first character in FindFirstChar, using a pattern the JIT can better optimize.
- Avoid using ToLower{Invariant} if we’re going to codegen out a call to a method that’s already case-insensitive, like char.IsDigit/IsWhiteSpace, and when using InvariantCulture, build the case-insensitivity into the lookup table so that we can similarly avoid the ToLower{Invariant} call on the fast path.
- Make the timeout check faster by using the unsigned rem.un instead of the signed rem
- Put back a couple of writes I accidentally removed in a previous PR, resulting in duplicating some unnecessary work in some cases.

cc: @danmosemsft, @eerhardt, @ViktorHofer, @GrabYourPitchforks 

|    Method |        Toolchain |              Options |       Mean |    Error | Ratio |
|---------- |----------------- |--------------------- |-----------:|---------:|------:|
|     Email | \new\corerun.exe | Ignor(...)riant [38] |   172.0 ns |  1.38 ns |  0.62 |
|     Email | \old\corerun.exe | Ignor(...)riant [38] |   275.8 ns |  4.14 ns |  1.00 |
|           |                  |                      |            |          |       |
|      Date | \new\corerun.exe | Ignor(...)riant [38] |   109.8 ns |  0.47 ns |  0.60 |
|      Date | \old\corerun.exe | Ignor(...)riant [38] |   182.8 ns |  0.97 ns |  1.00 |
|           |                  |                      |            |          |       |
|        IP | \new\corerun.exe | Ignor(...)riant [38] |   647.4 ns |  2.66 ns |  0.70 |
|        IP | \old\corerun.exe | Ignor(...)riant [38] |   930.4 ns |  5.17 ns |  1.00 |
|           |                  |                      |            |          |       |
|       Uri | \new\corerun.exe | Ignor(...)riant [38] |   110.7 ns |  0.85 ns |  0.66 |
|       Uri | \old\corerun.exe | Ignor(...)riant [38] |   168.8 ns |  1.29 ns |  1.00 |
|           |                  |                      |            |          |       |
| UriSpaces | \new\corerun.exe | Ignor(...)riant [38] |   300.4 ns |  2.27 ns |  0.34 |
| UriSpaces | \old\corerun.exe | Ignor(...)riant [38] |   886.7 ns |  6.39 ns |  1.00 |
|     Email | \new\corerun.exe | IgnoreCase, Compiled |   307.8 ns |  7.13 ns |  0.99 |
|     Email | \old\corerun.exe | IgnoreCase, Compiled |   312.6 ns |  3.55 ns |  1.00 |
|           |                  |                      |            |          |       |
|      Date | \new\corerun.exe | IgnoreCase, Compiled |   119.9 ns |  0.40 ns |  0.61 |
|      Date | \old\corerun.exe | IgnoreCase, Compiled |   196.7 ns |  1.52 ns |  1.00 |
|           |                  |                      |            |          |       |
|        IP | \new\corerun.exe | IgnoreCase, Compiled | 1,097.5 ns | 16.10 ns |  1.03 |
|        IP | \old\corerun.exe | IgnoreCase, Compiled | 1,067.8 ns |  8.51 ns |  1.00 |
|           |                  |                      |            |          |       |
|       Uri | \new\corerun.exe | IgnoreCase, Compiled |   193.1 ns |  1.07 ns |  0.99 |
|       Uri | \old\corerun.exe | IgnoreCase, Compiled |   194.5 ns |  1.85 ns |  1.00 |
|           |                  |                      |            |          |       |
| UriSpaces | \new\corerun.exe | IgnoreCase, Compiled |   966.6 ns | 11.87 ns |  1.00 |
| UriSpaces | \old\corerun.exe | IgnoreCase, Compiled |   966.2 ns |  4.03 ns |  1.00 |
|           |                  |                      |            |          |       |

```C#
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Running;
using System.Text.RegularExpressions;

public class Program
{
    static void Main(string[] args) => BenchmarkSwitcher.FromAssemblies(new[] { typeof(Program).Assembly }).Run(args);
}

public class Regexes
{
    [Params(
        RegexOptions.Compiled | RegexOptions.IgnoreCase,
        RegexOptions.Compiled | RegexOptions.IgnoreCase | RegexOptions.CultureInvariant)]
    public RegexOptions Options { get; set; }

    private Regex _email, _date, _ip, _uri;

    private const string EmailPattern = @"^([a-zA-Z0-9_\-\.]+)@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.)|(([a-zA-Z0-9\-]+\.)+))([a-zA-Z]{2,12}|[0-9]{1,3})(\]?)$";
    private const string DatePattern = @"\b(?<month>\d{1,2})/(?<day>\d{1,2})/(?<year>\d{2,4})\b";
    private const string IPPattern = @"(?:(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9])\.){3}(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9])";
    private const string UriPattern = @"[\w]+://[^/\s?#]+[^\s?#]+(?:\?[^\s#]*)?(?:#[^\s]*)?";
    private const string LotsOfSpaces = "                                                                                                                                                          ";

    [GlobalSetup]
    public void Setup()
    {
        _email = new Regex(EmailPattern, Options);
        _date = new Regex(DatePattern, Options);
        _ip = new Regex(IPPattern, Options);
        _uri = new Regex(UriPattern, Options);
    }

    [Benchmark] public void Email() => _email.IsMatch("yay.performance@dot.net");
    [Benchmark] public void Date() => _date.IsMatch("Today is 11/18/2019");
    [Benchmark] public void IP() => _ip.IsMatch("012.345.678.910");
    [Benchmark] public void Uri() => _uri.IsMatch("http://example.org");
    [Benchmark] public void UriSpaces() => _uri.IsMatch(LotsOfSpaces + "http://example.org");
}
```

stephentoub committed:
Remove unused RegexCompiler helpers

stephentoub committed:
Fix FindFirstChar to update runtextpos in the case of failure

I introduced this in a previous PR; by not updating the position in the case of failure, we end up potentially retrying some of the same text repeatedly.

stephentoub committed:
Improve codegen for FindFirstChar

For the common case of left-to-right, we can use a span to improve the code gen, which helps when there are multiple characters until the start of the regex.

stephentoub committed:
Optimize IgnoreCase and InvariantCulture codegen

We can improve our codegen for EmiCallCharInClass by avoiding the need to call char.ToLower{Invariant} in a variety of situations.  If the character class is one for which we already generate a check that handles case appropriately (e.g. using char.IsDigit for \d), there's no need to generate a ToLower call, regardless of culture.  And if we're using InvariantCulture, we can build the invariance into our generated lookup table, which means we only actually need the ToLowerInvariant call if the input isn't ASCII and we have to fall back to doing an actual CharInClass call.

stephentoub committed:
Fix stale comments

stephentoub committed:
Use rem.un instead of rem for timeout check

When a timeout is set, we try to avoid lots of CheckTimeout calls by only doing the call once every 2048 occurrences.  That check is currently done with a signed % operation; the codegen code for an unsigned % operation is better, and that's all we need.  We also don't need to explicitly compare to 0, and can just use brtrue on the rem.un result.

eerhardt commented on src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexInterpreter.cs:
```diff
@@ -484,6 +484,8 @@ protected override bool FindFirstChar()
                     }
                 }
+
+                runtextpos = runtextend;
```
How did you find these changes are needed? Is there any possible way of adding a test to guard it?

stephentoub replied:
I just happened to notice it was missing as compared to what had previously been there. Functionally, I don't think it matters; if FindFirstChar returns false in https://github.com/dotnet/runtime/blob/20ab2fa963934b5823e0b8afbba858169dd27169/src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexRunner.cs#L140, then we bump the current position by 1 and loop around to try again.  The issue would be performance, if we could have skipped more characters and are instead to repeatedly iterate by 1 instead of something larger.  I'm not sure of a good example to trigger it, though.

eerhardt reviewed (approved):
LGTM

lpereira commented on src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexCompiler.cs:
```diff
@@ -260,28 +263,14 @@ private void Ldc(int i)
         /// <summary>A macro for _ilg.Emit(OpCodes.Ret).</summary>
         private void Ret() => _ilg!.Emit(OpCodes.Ret);
 
         /// <summary>A macro for _ilg.Emit(OpCodes.Rem).</summary>
```
```suggestion
        /// <summary>A macro for _ilg.Emit(OpCodes.Rem_Un).</summary>
```

danmosemsft commented on src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexCompiler.cs:
```diff
@@ -40,6 +40,9 @@ internal abstract class RegexCompiler
         private static readonly MethodInfo s_charIsWhiteSpaceM = typeof(char).GetMethod("IsWhiteSpace", new Type[] { typeof(char) })!;
         private static readonly MethodInfo s_getcharM = typeof(string).GetMethod("get_Chars", new Type[] { typeof(int) })!;
+        private static readonly MethodInfo s_asspanM = typeof(MemoryExtensions).GetMethod("AsSpan", new Type[] { typeof(string), typeof(int), typeof(int) })!;
+        private static readonly MethodInfo s_spangetitem = typeof(ReadOnlySpan<char>).GetMethod("get_Item", new Type[] { typeof(int) })!;
```
"s_spangetitemM" ?

danmosemsft replied:
Nit, caps aren’t consistent in these names either. Might be clearer to retain caps And underscores

danmosemsft commented on src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexInterpreter.cs:
```diff
@@ -1045,7 +1049,7 @@ protected override void Go()
+                                // Check the timeout every 2048th iteration. The additional if check
                                 // in every iteration can be neglected as the cost of the CharInClass
                                 // check is many times higher.
                                 if ((uint)c % LoopTimeoutCheckCount == 0)
```
Hopefully the JIT can determine that it can just do a bit mask to find N % 2048 and not an expensive mod operation. If not, we could do it explictly if it matters.

mikedn replied:
> Hopefully the JIT can determine that it can just do a bit mask to find N % 2048 and not an expensive mod operation

Yes. Especially that it's unsigned mod.

danmosemsft replied:
Yes it can or yes it ideally would?

mikedn replied:
Yes it can. It generates an `and` instruction:
```asm
       25FF070000           and      eax, 0x7FF
```

danmosemsft commented on src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexCompiler.cs:
```diff
@@ -675,10 +667,13 @@ private void InitLocalCultureInfo()
         }
 
+        /// <summary>Whether ToLower operations should be performed with the invariant culture as opposed to the one in <see cref="_cultureV"/>.</summary>
+        private bool UseToLowerInvariant => _cultureV == null || _options.HasFlag(RegexOptions.CultureInvariant);
```
HasFlag is intrinsic these days, fast enough you don't need to cache this?

stephentoub replied:
Yes:
https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIAYACY8gOgBUYEMWAlGAcwCuAG2xQAoggAOsXLgCWEAHa4A3DRrEAzE1IMAwgwDeNBmaY7gECMIYBZcgAo+/TgHkpGRSoYRP33ABKBgBeAD5ff2VcFgAJPAAxUX5nAXcolRZ9EQxBWABJJQA3MXlsJQxA9WpzCwYrG3tSVNcEDy9oyI6VYPCGRz9u3AYAMgYXdKGsnLyYQpKoMorggEIQhjpqgF8gA
But I could be explicit about it if you'd prefer.

danmoseley reviewed (approved):
It's hard to review well but we should have pretty good code coverage.

danmosemsft commented on src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexCompiler.cs:
```diff
@@ -2159,18 +2214,18 @@ private void GenerateOneCode()
                     }
                     else
                     {
+                        if (IsCi())
```
Nit, it would be nice if they'd named this `IsCaseInsensitive()`

stephentoub committed:
Fix stylistic issues cited in code reviews

- Renamed all RegexCompiler static fields to be consistently named after the fields they represent
- Renamed some methods that were using acronyms
- Renamed LocalBuilers to be named consistently after the fields they mirror, and also to use consistent suffixes
- Fixed some comments
- Removed use of Enum.HasFlag to be consistent with bit flags checks elsewhere

stephentoub commented:
@danmosemsft, I pushed a style-only commit to address the feedback (mostly changing things I didn't actually add in this PR).
